Proper installer

hansfilipelo opened the pull request:
This PR adds:
- Proper installer and config files for wpManySiteAdmin
- Crontab file (installed with the rest)
- A main executable, wpma, which is used as main interface
- A default config
- More config options (enable/disable HTTPS and AD-plugin)

Rovanion committed:
Renamed project and most scripts.

The scripts according to the convention of action/component.

The project itself to Wordpress Mass Administration, or wp-mass-admin for short.

Commit:
Merge remote-tracking branch 'origin/master' into change-names

hansfilipelo committed:
Follow new naming convention

hansfilipelo committed:
Corrected bugs which led to older filenames

Rovanion committed:
Added humane error message on conf file missing.

Rovanion committed:
Corrected cd paths in install instructions.

hansfilipelo committed:
First commit of new installer

hansfilipelo committed:
Installer works as intended. Remains to fix starter script in order to test all functions from new locations.

hansfilipelo committed:
Simple starter script for wpma

hansfilipelo committed:
wpma checks for running as root or not

hansfilipelo committed:
New wpma-starter can install a site without HTTPS, nothing else tested

hansfilipelo committed:
New wpma can delete a site using HTTP

hansfilipelo committed:
Install and delete of HTTPS-site tested and is working

hansfilipelo committed:
Intermediate commit

hansfilipelo committed:
Installer writes username to config file, this one is for you rovanion

hansfilipelo committed:
Cron does not accept this crontab - no idea why.

hansfilipelo committed:
Crontab works

hansfilipelo committed:
Crontab works

hansfilipelo committed:
Does not install AD-plugin by default

hansfilipelo committed:
When deleting a site it is removed from the config file

hansfilipelo committed:
Added common folder

hansfilipelo committed:
Writer was missing from earlier

hansfilipelo committed:
Update docs for new version

Rovanion commented on README.md:
```diff
@@ -1,15 +1,17 @@
-Scripts for simplified installation, update, monitoring and deletion of Wordpress sites. Each site is after installation automatically upgraded through minor updates and scanned for vulnerabilities using wp-scan and ssl-labs-scan.
+Scripts for simplified installation, maintenance, monitoring and deletion of Wordpress (WP) sites. Each site is after installation automatically upgraded through minor updates and scanned for vulnerabilities using wp-scan and ssl-labs-scan. They're also set up with HTTPS-certificates through Let's Encrypt.
 
 Note: The default installation installs the Active Directory (AD) plugin for Wordpress since AD is commonly used at LiU for user data. This readme does NOT cover AD-plugin configuration since that's specific to each domain.
```
Visst kan vi ändra det här beteendet nu och så även README:n?

hansfilipelo replied:
Default behaviour concerning AD is changed - just missed it in README.

Rovanion commented on common/conf.example:
```diff
@@ -6,12 +6,19 @@ adminMail=admin@example.com
-userNames="org1 org2"
+userNames=""
+
+# Use letsencrypt for HTTPS? If so, DNS records need to be accurate before installing!
```
Accurate in what way?

Perhaps we can write that DNS records must point to the server running the script in order for installation to work?

hansfilipelo replied:
Accurate in that way. 

It _almost_ says exact that - of you'd like to formulate it differently please do :)

Rovanion replied:
Well, that's what I did, perhaps I can express it more succintly: `s/accurate/point to the server running this script/`.

hansfilipelo replied:
Changed the scentence.

Rovanion commented on delete/README.md:
```diff
@@ -1,7 +1,7 @@
 
 <pre>
-sudo ./deleteWPInstall.bash userName FQDN
+sudo ./delete/wordpress-install userName FQDN
```
Shouldn't this line be `sudo wpma delete userName FQDN`.

Rovanion commented on delete/wp-site:
```diff
@@ -9,8 +9,11 @@ fi
 
+if [[ -e /usr/local/etc/wp-mass-admin/wpma.conf ]]; then
+    source /usr/local/etc/wp-mass-admin/wpma.conf
+fi
```
Should we not reach some failure state if there is no conf? Perhaps we should also check in `/etc/wp-mass-admin` before bailing?

hansfilipelo replied:
Do we - even if default conf exists?

Rovanion replied:
Hmm, it's not a clear cut case now that I read the code again. I'd have to ask some questions first:

1. Is there anything the default conf in a clean installation?
2. Which conf file is automatically written to?
3. IFF the automatic addition of site names is to the non-default-conf, does it make sense to attempt to delete the given site since it's reasonable to assume that there are none?

hansfilipelo replied:
1. Default conf is same as the empty "regular" right now
2. Non-default
3. Probably not. Changing behaviour.

hansfilipelo committed:
Changed old documentation

Rovanion commented on delete/wp-site:
```diff
@@ -28,18 +31,27 @@ rm -f /etc/nginx/sites-enabled/"$FQDN"
+	
+	# Remove config for domain
+	rm -f "$letsencryptFolder"/"$FQDN".cli.ini
+fi
```
There should be some else statement here if only to inform the user that no Let's Encrypt conf was found for the site and that we assume that no certificates need to be revoked.

hansfilipelo replied:
Perhaps more appropriate - prompt if there is one and ask whether user wants to revoke it or not?

Rovanion commented on delete/write-conf:
```diff
@@ -0,0 +1,46 @@
+	done
+	echo "\"" >> $tmp_file
+    fi
+done < $config_file
```
Hmm, I'm having trouble grokking the above script. What does it do differently that can't simply be done with:

```sed
sed 's/userNames="(.*)"/userNames="\1 $newUser"/'
```

Rovanion commented on delete/write-conf:
```diff
@@ -0,0 +1,46 @@
+while read line
+do
+    echo $line | grep "userNames" >> /dev/null
+    if [ $? != 0 ]
```
Line 12 and 13 should be combined as:

```
if echo $line | grep "userNames" >> /dev/null; then do something; fi
```

hansfilipelo replied:
No this is incorrect. Watch how $? is not equal to 0.

Rovanion commented on delete/write-conf:
```diff
@@ -0,0 +1,46 @@
+cat $tmp_file > $config_file
+rm $tmp_file
+
+
```
Trying again to grokk the above script. I beg forgiveness for the spammy emails. I obviously hadn't understood the purpose of the script last time. Let me try again by asking:

Why can't the script simply be replaced with:

```sed
sed -r "s/userNames=\"(.*)$old_user(.*)\"/userNames=\"\1\2\"/""
```

Rovanion commented on install/no-https:
```diff
@@ -0,0 +1,82 @@
+		fastcgi_param	SCRIPT_FILENAME	\$document_root/\$fastcgi_script_name;
+		include fastcgi_params;
+	}
+}
```
Everything between line 30 and here except for the `error_log`, `access_log` and `fastcgi_pass` statements can be refactored out to shared snippets included back in with the `include` statement.

The `include` statement seems to be the only way to achieve this, there seems to be no way to do parameterized includes or function calls in nginx-confs.

hansfilipelo replied:
This is fixed in the most recent commit (15428cb).

Rovanion commented on install/wp-site:
```diff
@@ -36,20 +46,27 @@ chown -R $userName:www-data $baseDir
+    su $username -c "cd $installDir
 wp plugin install active-directory-integration
 wp plugin activate active-directory-integration"
+fi
```
If this is all we do, why don't we let the user provide an array of plugins to be installed instead of hard-coding it? We could have active-directory-integration as an example in the manual.

Rovanion commented on install/write-conf:
```diff
@@ -0,0 +1,31 @@
+
+cat $tmp_file > $config_file
+rm $tmp_file
+
```
Same question about this script as with the other write-conf one. Couldn't it be replaced with:

```sed
sed -r "s/userNames=\"(.*)\"/userNames=\"\1 $new_user\"/""
```

hansfilipelo replied:
Not strong with my regex - probably but please try?

Rovanion commented on install/wp-site:
```diff
@@ -36,20 +46,27 @@ chown -R $userName:www-data $baseDir
+echo "Installing AD-plugin"
+if [[ $active_directory == "yes" ]]
+then
+    su $username -c "cd $installDir
```
Found a typo, fixed with: `s/$username/$userName`.

Rovanion commented:
Finally I have some questions on the subject of code style; because I do think we have to be consequential, no matter which style we choose to follow.


In earlier code we've used

```bash
if [[ "$useHttps" == "yes" ]]; then
    ./use-https $userName $FQDN
fi
```

while in this pull request you used


if [[ "$useHttps" == "yes" ]]
then
    ./use-https $userName $FQDN
fi


And in parts of the new files you've written you used snake_casing like `new_user="$1"` from `install/write-conf` while you've kept using camelCase in others.

In both cases we should decide on one style and stick to it.

Rovanion commented:
We must also decide whether or not to use tab characters or spaces for indentation. Liberal mixing can now be blatantly be seen on our lines, absolutely [unacceptable](https://www.youtube.com/watch?v=aaSRYecKaqc)!

Rovanion commented on install/nginx:
```diff
@@ -0,0 +1,29 @@
+else
+	./no-https $userName $FQDN
+fi
+
```
To me it looks like the user is only ever able to choose whether or not to use HTTPS for an install in the global config file and not on an invocation basis. I would say that in our situation in CYD we want the default to be HTTPS but with the option of opting out for the setup of some specific sites.

hansfilipelo replied:
It is possible to change the config before running. 

However I agree with you that for the future the command line should offer more options, however I feel more like doing that in a proper python starter (replace the wpma bash script starter with one written in Python). I feel like this together with the installer refractor is a really large scope and perhaps we can delay these features to antoher PR.

Rovanion commented on install/nginx:
```diff
@@ -0,0 +1,29 @@
+
+userName=$1
+FQDN=$2
+configFile=/etc/nginx/sites-available/$FQDN
```
Here we define the path to the config file, but we never load it.

Rovanion commented on install/no-https:
```diff
@@ -0,0 +1,82 @@
+
+# Create logfolder+configfile
+mkdir -p /var/log/nginx/$FQDN
+touch $configFile
```
Why is this touch line needed? Tried running the cat command below without it and it seemed to create the file just fine without it.

Rovanion committed:
Corrected quotation of all variables in all scripts.

Written with the help of shellcheck: https://github.com/koalaman/shellcheck/

This commit has not been tested, but should contain no semantic differences assuming normal use of the program.

Rovanion commented:
I've been at it for four hours so I haven't tested the functionality of the latest commit. Also note that the latest commit made GitHub think that a lot of the above comments are out of date while most of them remain unadressed.

hansfilipelo committed:
Made config writers work again. Added include file for nginx

hansfilipelo commented:
Regarding camelCase vs under_score i prefer using under_score now, but when i started the project I was of another opinion. I'll see tot that things are in better shape. 

Regarding tabs: My former wasn't set to replace tabs with spaces. Each time I open a file now it will be spaces only. Please let me know which files are tabs in.

hansfilipelo committed:
Change behaviour on delete so that user needs to be in config in order for script to run (in order to somewhat mitigate miss-spelling)

hansfilipelo committed:
Changed camelCaseSyntax to under_score_syntax

hansfilipelo committed:
Syntax change

hansfilipelo committed:
Syntax change

hansfilipelo committed:
Syntax change in readme

hansfilipelo commented:
I've wither answered all comments or changed according to comments. Ready for more testing and opinions/suggestions.

hansfilipelo commented:
When this is merged I suggest we change the repository name.

hansfilipelo committed:
More syntax fix

hansfilipelo committed:
Merge branch 'proper_install' of github.com:CYD-poolen/wpManySiteAdmin into proper_install

hansfilipelo committed:
Stricter permissions for www-data

hansfilipelo committed:
Fix for folder not existing when setting permissions
